src/lib.rs

```diff
@@ -0,0 +1,631 @@
+            }
+            Ok(UMessageType::UmessageTypeResponse) => {
+                Validators::Response
+                    .validator()
+                    .validate(&attributes)
+                    .map_err(|_| {
+                        UStatus::fail_with_code(
+                            UCode::InvalidArgument,
+                            "Wrong Response UAttributes",
+                        )
+                    })?;
+                self.send_response(&zenoh_key, payload, attributes).await
+            }
+            _ => Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Wrong Message type in UAttributes",
+            )),
+        }
+    }
+
+    async fn register_listener(
+        &self,
+        topic: UUri,
+        listener: Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>,
+    ) -> Result<String, UStatus> {
+        // Do the validation
+        UriValidator::validate(&topic)
+            .map_err(|_| UStatus::fail_with_code(UCode::InvalidArgument, "Invalid topic"))?;
+
+        // Get Zenoh key
+        let zenoh_key = ULinkZenoh::to_zenoh_key_string(&topic)?;
+        // Generate listener string for users to delete
+        let hashmap_key = format!(
+            "{}_{:X}",
+            zenoh_key,
+            self.callback_counter.fetch_add(1, Ordering::SeqCst)
+        );
+
+        // Setup callback
+        let callback = move |sample: Sample| {
+            // Create UAttribute
+            let Some(attachment) = sample.attachment() else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to get attachment",
+                )));
+                return;
+            };
+            let Some(attribute) = attachment.get(&"uattributes".as_bytes()) else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to get uattributes",
+                )));
+                return;
+            };
+            let Ok(u_attribute) = Message::decode(&*attribute) else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to decode attribute",
+                )));
+                return;
+            };
+            // Create UPayload
+            let Ok(encoding) = sample.encoding.suffix().parse::<i32>() else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "Unable to get payload encoding",
+                )));
+                return;
+            };
+            let u_payload = UPayload {
+                length: Some(0),
+                format: encoding,
+                data: Some(Data::Value(sample.payload.contiguous().to_vec())),
+            };
+            // Create UMessage
+            let msg = UMessage {
+                source: Some(topic.clone()),
```

Comment:
In a similar fashion to @stevenhartley's [comment over here](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2/files#r1463406504), we'd need to always get the actual source.

Why? In support of a "special mode" for `UTransport::register_listener()` that can support listening to all remotely addressed CEs that the uStreamer would need.